### DISCUSSION TITLE
Using graphql-java-codegen with graphql-java/java-dataloader

### DISCUSSION THREAD

**Discussion by ddskolla** [kobylynskyi/graphql-java-codegen#792] [Q&A] [answered]
Hi,

I am attempting to use the dataloader pattern(https://github.com/graphql-java/java-dataloader) in our graphql api.
To get the maximum out of the data loader pattern it is recommended to use an async flow to return data (using CompletableFuture etc). At the moment, the resolver interfaces generated by the codegen tool, returns the standard POJO objects. Like : 
```
public interface UserQueryResolver extends graphql.kickstart.tools.GraphQLQueryResolver {

    User user(String id, graphql.schema.DataFetchingEnvironment env) throws Exception;

}
```

Is there any way that you would recommend to generate the resolvers that is suitable for this use-case ?

Ideally, the interface should be generated like
```
)
public interface UserQueryResolver extends graphql.kickstart.tools.GraphQLQueryResolver {

    CompletableFuture<UserTO> user(String id, graphql.schema.DataFetchingEnvironment env) throws Exception;

}
```

Any advice on using graphql-java-tools , graphql-java-codegen along with java-dataloader would be appreciated as there doesn't seem to be a standard pattern for this.

**Comment by kobylynskyi** [comment, marked as answer]
@ddskolla please consider using `apiReturnType`:
```xml
<apiReturnType>java.util.concurrent.CompletableFuture</apiReturnType>
```
or 
```groovy
apiReturnType = "java.util.concurrent.CompletableFuture"
```
